test/sql/test_types.py

```diff
@@ -383,6 +384,28 @@ def load_dialect_impl(self, dialect):
         assert t1.dialect_impl(d)._type_affinity is postgresql.UUID
 
 
+class AsGenericTest(fixtures.TestBase):
+    @testing.combinations(
+        (String(), String()),
+        (VARCHAR(length=100), String(length=100)),
+        (NVARCHAR(length=100), Unicode(length=100)),
+        (DATE(), Date()),
+        (pg.JSON(), sa.JSON()),
+        (pg.ARRAY(sa.String), sa.ARRAY(sa.String)),
```

Comment:
**mike bayer** (zzzeek) wrote:

trickier types to add here:  types.Enum, mysql.ENUM, postgresql.ENUM, postgresql.INTERVAL

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/2375

Reply:
I have `Enum` type tests now passing, but postgres.INTERVAL is not as straightforward.  For types such as INTERVAL that are truly database specific and don't have a natural generic analog, does it make sense for `as_generic` to not be supported? Not many great options in the `__mro__`:

```
In [39]: INTERVAL().__class__.__mro__
Out[39]: 
(sqlalchemy.dialects.postgresql.base.INTERVAL,
 sqlalchemy.sql.type_api.NativeForEmulated,
 sqlalchemy.sql.sqltypes._AbstractInterval,
 sqlalchemy.sql.sqltypes._LookupExpressionAdapter,
 sqlalchemy.sql.type_api.TypeEngine,
 sqlalchemy.sql.visitors.Traversible,
 object)
```

For this

Reply:
well we at least have to start with some kind of "defined" behavior, even if that is NotImplementedError.  However there is a generic Interval datatype: https://docs.sqlalchemy.org/en/13/core/type_basics.html?highlight=interval#sqlalchemy.types.Interval     if I was reflecting a PG database with INTERVAL and I wanted to create that table in SQLite I'd want to use the generic Interval datatype , no? it stores the data differently but if one were using SQLAlchemy to pull and move the data it would be converting it.

Reply:
for types like those, they can implement their own special as_generic() method that's hardcoded to what they return.  we don't have to rely upon the heuristic for every type, it's just to catch the vast majority of them.

Reply:
Makes sense - I did notice this: https://github.com/sqlalchemy/sqlalchemy/blob/206afac627ee79ff2ce7b1305615e471fb34503b/lib/sqlalchemy/dialects/postgresql/base.py#L1434

Which made me think how if we were refactoring the existing _type_affinity instead of writing a new _generic_type_affinity, this bit of code above would automatically have handled this for us

Reply:
So just want to confirm that we do want to introduce _generic_type_affinity instead of just updating the existing _type_affinity?

Reply:
this can be for _type_affinity if we're only in the 1.4 series as long as no existing uses of type_affinity break, which I believe is mostly if not all in the test suite.

Reply:
On closer consideration that might be a whole can of worms. It breaks some tests. And furthermore, they really are doing different things.  `_type_affininity` finds the deepest class in the MRO thats not TypeEngine. Whereas `_generic_type_affinity` is finding the shallowest class in the MRO that will pass the criteria for a generic.  It wants the generic that most closely resembles the native type, so it stops as soon as it finds one that is considered a generic.

Reply:
@zzzeek As it is, as_generic returns the NullType when it can't find a good type to return.  This makes sure that as_generic always "succeeds" and has a defined behavior. Just wanted to call that out - is it preferrable for as_generic to fail more loudly with a NotImplementedError when the as_generic heuristic fails?

Reply:
I'd prefer if that can raise by default; we can also have an option "allow_nulltype" or similar that will return NullType if the user wants it to succeed in any case.   does that work?

Reply:
Works for me - finishing this up today